### ISSUE TITLE
win_package: does not access user_name + user_password for samba share, local authentication

### ISSUE THREAD

**Issue by dreamcat4** [ansible/ansible-modules-extras#1140]
##### ISSUE TYPE

Feature Idea
##### COMPONENT NAME

win_package module
##### ANSIBLE VERSION

N/A
##### SUMMARY

Hi.
I'm trying to install a package from network share path. However it does not connect / authenticate to the samba share.

The error returned looks likt a kerberos error. But I don't use kerberos connectoin method. Only local auth. So maybe your module is assuming some things.

``` sh
TASK: [base | Install Eraser] ************************************************* 
<ga-z170x-g3> REMOTE_MODULE win_package name=Eraser user_name=***** path='\\smb.ansible\ansible\windows\freeware\Eraser 6.2.0.2970.exe' arguments=/S user_password=***** product_id=folaskdjlj
failed: [ga-z170x-g3] => {"changed": false, "failed": true, "name": "Eraser"}
msg: System.ComponentModel.Win32Exception (0x80004005): A specified logon session does not exist. It may already have been terminated

FATAL: all hosts have already failed -- aborting
```

**Comment by dreamcat4** [comment]
@trondhindenes Hi there. I was hoping to somehow fix or improve your powershell code in this module. To get samba share working on local accounts. But so-far I really can't see any specific point where there might be an issue with your code.

Could this failure be because ansible insists on running each and every winrm cmds seperately as `'PowerShell' ['-NoProfile', '-NonInteractive',` ?

^^ Why these flags! :/. It seems to cause issues for many software installers, who assume a login session. It seems like a possible reason for my error message here ^^ about login session. i.e. that it can't remember and credentials. "A specified logon session does not exist..." and so on.

I've tried doing `net use` in a separate task immediately prior to `win_package` to register the login credentials:

``` yaml
- name: Pass username and password credentials for connecting to local samba share
  raw: net use \\smb.share.lan /user:mywinhost\username password /persistent:yes
```

Unfortunately it don't make any difference / no effect. How can we tackle this?

Could we try putting that cmd into your module? So it can execute directly before trying to connect to the samba share? Perhaps it would remember the login credentials.

Thing is your module appears to assume windows domain controlled and/or do nice powershell bits which are currently beyond my understanding. Don't want to muck that up for anybody else here. Or add unnecessary extra configuration flags if they are not really needed.

**Comment by dreamcat4** [comment]
And here's another one:

``` sh
failed: [ga-z170x-g3] => {"changed": false, "failed": true, "name": "Eraser"}
msg: System.InvalidOperationException: The process \\smb.ansible\ansible\windows\freeware\Eraser 6.2.0.2970.exe could not be started ---> System.Management.Automation.MethodInvocationException: Exception calling "Start" with "0" argument(s): "Access is denied" ---> System.ComponentModel.Win32Exception: Access is denied
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at CallSite.Target(Closure , CallSite , Object )
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   --- End of inner exception stack trace ---

FATAL: all hosts have already failed -- aborting
```

It kindda feels like the same UAC error I get whenever I try executing any `.exe` or other programs from a samba share. Although it isn't UAC, maybe we would have better luck if the file is first to be copied over to the target windows machine, and then executed from there. Like the HTTP path method.

Well "sort of". Because it isn't consistent or reliable (winndows) for the times I get even that far. For example, not changing anything else. I can re-run the sam cmd 2 more times. Then (for no aparrent reason, and nothing in ansible was changed):

```
failed: [ga-z170x-g3] => {"changed": false, "failed": true, "name": "Eraser"}
msg: System.ComponentModel.Win32Exception (0x80004005): A specified logon session does not exist. It may already have been terminated

FATAL: all hosts have already failed -- aborting
```

What would be cool is some kind of a `-trace` mode to print out the execution lines of your module `.ps1` script? You know, like the equivalent of `set -x` on linux. That would be helpful to me. Any way we can get such debugging infos come back to us over the winrm?

Many thanks for any help / assistance. I still can't follow / understand your powershell functions in this module. Not an expert.

**Comment by trondhindenes** [comment]
Good find. I'll take a look and report back. As for debugging, you can just set `export ANSIBLE_KEEP_REMOTE_FILES=1` and the powershell script will persist on the target node so that you can debug it. (appdata\temp folder of the user executing the script)

**Comment by dreamcat4** [comment]
Thanks @trondhindenes for the debugging tip I appreciate that.

When you investigate, it may turn out these weird error messages are down to credentials limitations in windows itself. What struck me was the inconsistency of these different errors between runs.

So I've setup a local http server in the meantime. Which is working very reliably.

**Comment by trondhindenes** [comment]
ok great. The credentials passed are used  in line 647 in the module, basically creating a psdrive using new-psdrive with the credential as a parameter (among others). It would be interesting to test wether you're able to construct a psdrive from your windows box to the samba share using those same credentials.

**Comment by dreamcat4** [comment]
Ah interesting. I followed your suggestion and executed the kept downloaded script locally. And from inside an admin powershell session, of logged in user account. But it gets the same error message:

https://gist.github.com/dreamcat4/497bbd61889a8716c187

There is only 1 samba username I have ever used as a login for that samba share (`ansible`).

http://powershell.com/cs/forums/t/22452.aspx

Taking the ^^ above linked advice and putting the IP address in the `New-PsDrive` command was the only thing that worked. Windows 10 here.

You can see here how I constructed the PSDrive cmds:

``` powershell
PS>$pass="FooBoo"|ConvertTo-SecureString -AsPlainText -Force
PS>$Cred = New-Object System.Management.Automation.PsCredential('user@domain',$pass)
PS>New-PSDrive -name j -Root \\server\share -Credential $cred -PSProvider filesystem

Name           Used (GB)     Free (GB) Provider      Root
----           ---------     --------- --------      ----
j                                      FileSystem    \\server\share
```

Source: [stackoverflow](http://stackoverflow.com/questions/13818751/how-can-i-create-a-new-psdrive-by-using-a-password-that-is-not-systemsecuredstri/13819230#13819230)

My interactive session (scroll to bottom):

https://gist.github.com/dreamcat4/159485d2b0621bac0352

Recommendation:

Maybe consider going back to the tried and tested `net use` if possible.

``` sh
 net use /?
The syntax of this command is:

NET USE
[devicename | *] [\\computername\sharename[\volume] [password | *]]
        [/USER:[domainname\]username]
        [/USER:[dotted domain name\]username]
        [/USER:[username@dotted domain name]
        [/SMARTCARD]
        [/SAVECRED]
        [[/DELETE] | [/PERSISTENT:{YES | NO}]]

NET USE {devicename | *} [password | *] /HOME

NET USE [/PERSISTENT:{YES | NO}]
```

Had enough of this now. Going back to my local http fileserver (nginx).

**Comment by dreamcat4** [comment]
There is also something called "New-CimSession" too. Although I doubt it will solve such issue, here is a link for completeness anyway:

https://technet.microsoft.com/en-us/library/jj590760%28v=wps.630%29.aspx

**Comment by trondhindenes** [comment]
I'm afraid CimSession is not related here. The error you're getting is interesting, indicating that you already have an SMB session against the file server using a different user name (Windows doesn't support that). I might have to stand up a samba server to take a closer look at this.

**Comment by trondhindenes** [comment]
Btw, the "sessions" it complains against is regarding the target (samba) server, not the share. So if you already have an open session against another share on the same server with a different username (or anonymous) then it's never going to work. That is just how Windows works. But I'll try and get this tested anyhow.

**Comment by nitzmahone** [comment]
To expand a bit on what @trondhindenes has pointed out, MS makes it _really hard_ to seed additional credentials into the cache over WinRM. Even net use has issues on newer versions of Windows, and the Powershell credential cache management cmdlets are deliberately crippled over WinRM. We've got some connections inside Microsoft now that might be able to explain that rationale. The best thing we could do here would be adding CredSSP support (which is tentatively on the roadmap for 2.2)- that's the only "supported" way to do this.

**Comment by sparkvilla** [comment]
@nitzmahone I would be interested to know how the CredSSP support could be integrated into Winrm. I am trying to install a chocolatey package located in a samba share on a windows client through Ansible. I think I am unable to do that  because of the so-called "second hop" issue (see https://github.com/ansible/ansible-modules-extras/issues/1886 ). Any idea how could I solve this problem using CredSSP or something else?

**Comment by dreamcat4** [comment]
Indeed I would also like to know too. If indeed any way to use samba for hosting files for an ansible run is possible.

My workaround until then is to run a simple nginx container on my docker server on same LAN. It just serves up any such installer files via http. Then can be used in `win_get_url` or `win_package` / similar etc within the windows playbooks. So much less hassle than samba is. No credentials needed.

**Comment by trondhindenes** [comment]
I haven't gotten around to testing this yet, but it could be that a WinRM session which is Kerberos-based is unable to connect to a samba/smb share using any other credential than the one used in the Kerberos session. On ansible this can simply be tested by switching from Kerberos to basic auth for winrm traffic.

As for the double-hop problem, there are 2 options: 1: Use something else than Kerberos, or 2: Use "Trust for delegation" (https://technet.microsoft.com/en-us/library/cc739740%28v=ws.10%29.aspx?f=255&MSPPError=-2147217396)

**Comment by sparkvilla** [comment]
@trondhindenes I think I am not using kerberos autentication because I haven't installed it on my host. How can I check for which auth am I using and how to switch from one to another? I am quite new to ansible

**Comment by trondhindenes** [comment]
Good question. Well if you run klist on your ansible node and nothing is returned (or an error is returned because of missing packages) you are definetely not using Kerberos. 

Ansible has some auto-sensing for which auth type to use, based on how the username is specified. More info at http://docs.ansible.com/ansible/intro_windows.html#configuring-kerberos

**Comment by nitzmahone** [comment]
If you're using kerberos, we've added kerberos delegation support to a new version of pywinrm (that should be available roughly around 2.1 release timeframe)- with that turned on, your credentials can make a transparent second-hop to an SMB share. Once pywinrm >=0.1.2 has shipped (it might be 0.2.0 or 1.0.0, not sure of version yet), you can just add `ansible_winrm_kerberos_delegation=true` to your target host's inventory vars and it should work. This should also work on any version of Ansible back to 2.0 with the new pywinrm installed.

**Comment by ansibot** [comment]
@trondhindenes, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jhawkesworth** [comment]
I think that the 'fix' for this issue is to document the available options, which seem to be either use Kerberos auth delegation or workaround using an http server to fetch files from.

**Comment by trondhindenes** [comment]
@jhawkesworth I tend to agree. The second-hop issue keeps coming up in the forums as well, so maybe someone should just sit down and explain what works and what doesn't. Then module help could just link that page.

**Comment by ansibot** [comment]
@trondhindenes, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jhawkesworth** [comment]
needs_contributor (to write docs)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by dagwieers** [comment]
Ok, after reviewing this issue, I think we can close it. The documentation has been improved in the meantime with the various authentication options and their characteristics. The support code has been improved a lot recently and if any ill-effects are encountered, please open a new ticket at https://github.com/ansible/ansible/issues

Thanks !
